distributed/client.py

```diff
@@ -753,6 +761,7 @@ class Client(SyncMethodMixin):
 
     _default_event_handlers = {"print": _handle_print, "warn": _handle_warn}
 
+    _set_config_stack = contextlib.ExitStack()
```

Comment:
this is a ClassVar and so there's a global ExitStack at `distributed.client.Client._set_config_stack` and ExitStack isn't threadsafe

Reply:
maybe something like this?

```python
class _ConfigureDask:
    def __init__(self):
        self._rlock = threading.RLock()
        self._count = 0

    def __enter__(self):
        with self._rlock:
            self._count += 1
            if self._count == 1:
                self._set = dask.config.set(scheduler="dask.distributed", shuffle="tasks")

    def __exit__(self, *exc_info):
        with self._rlock:
            self._count += 1
            if self._count == 0:
                self._set.__exit__(None, None, None)
                del self._set

_configure_dask = _ConfigureDask()
```

Reply:
or maybe something like this:


```python
class _GlobalClientManager:
    def __init__(self):
        self._lock = threading.RLock()
        self._global_clients: weakref.WeakValueDictionary[
            int, Client
        ] = weakref.WeakValueDictionary()
        self._global_client_index = 0
        self._set = None

    def _get_global_client(self) -> Client | None:
        with self._lock:
            L = sorted(self._global_clients, reverse=True)
            for k in L:
                c = _global_clients[k]
                if c.status != "closed":
                    return c
                else:
                    del _global_clients[k]
            return None

    def _set_global_client(self, c: Client) -> None:
        with self._lock:
            if not self._set:
                self._set = dask.config.set(
                    scheduler="dask.distributed", shuffle="tasks"
                )
            self._global_clients[self._global_client_index] = c
            self._global_client_index += 1

    def _del_global_client(self, c: Client) -> None:
        with self._lock:
            for k in list(self._global_clients):
                try:
                    if _global_clients[k] is c:
                        del _global_clients[k]
                except KeyError:  # pragma: no cover
                    pass

            if not self._global_clients:
                self._set.__exit__(None, None, None)
                self._set = None


_global_client_manager = _GlobalClientManager()
_set_global_client = _global_client_manager._set_global_client
_get_global_client = _global_client_manager._get_global_client
_del_global_client = _global_client_manager._del_global_client
```

and it would fix https://github.com/dask/distributed/issues/5772 also

Reply:
I'm not sure what our appetite for metaclasses in the code base is, but I've personally liked using [Multiton's](https://en.wikipedia.org/wiki/Multiton_pattern) for heavyweight resource objects which should be unique per-process for the arguments used to create it.  

```python
import weakref
from threading import Lock

class ClientMultiton(type):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.__cache = weakref.WeakValueDictionary()
        self.__lock = Lock()

    def __call__(cls, *args, **kw):
        key = args + (frozenset(kw.items()),) if kw else args

        with cls.__lock:
            try:
                # Obtain existing instance for this key
                instance = cls.__cache[key]
            except KeyError:
                pass
            else:
                # Return instance if it's not closed
                if not instance.closed:
                    return instance

                # Delete the closed cache entry
                # and proceed with recreating the instance
                # for this key
                del cls.__cache[key]

            cls.__cache[key] = instance = type.__call__(cls, *args, **kw)
            return instance


class Client(metaclass=ClientMultiton):
    def __init__(self, address=None, timeout=100):
        self.address = address
        self.timeout = timeout
        self._closed = False
        self._lock = Lock()

    @property
    def closed(self):
        # Dangerous if self._closed can transition from False to True
        with self._lock:  
            return self._closed

    def close(self):
        with self._lock:
            if self._closed:
                return

            print(f"Closing {self.address}")
            self._closed = True

    def __del__(self):
        self.close()

if __name__ == "__main__":
    def inner():
        c1 = Client(address="tcp://172.0.0.1")
        assert Client(address="tcp://172.0.0.1") is c1
        assert Client(address="tcp://172.0.0.1") is Client(address="tcp://172.0.0.1")

        addresses = list(sorted(c.address for c in Client._ClientMultiton__cache.values()))
        assert addresses == ["tcp://172.0.0.1"]

        c2 = Client(address="tcp://172.0.0.2")
        assert c2 is not c1
        addresses = list(sorted(c.address for c in Client._ClientMultiton__cache.values()))
        assert addresses == ["tcp://172.0.0.1", "tcp://172.0.0.2"]

        c1.close()
        addresses = list(sorted(c.address for c in Client._ClientMultiton__cache.values()))
        assert addresses == ["tcp://172.0.0.1", "tcp://172.0.0.2"]

        c3 = Client(address="tcp://172.0.0.1")
        assert c3 is not c1
        assert not c3.closed
        addresses = list(sorted(c.address for c in Client._ClientMultiton__cache.values()))
        assert addresses == ["tcp://172.0.0.1", "tcp://172.0.0.2"]
    
    inner()
    print("Done")
    assert len(Client._ClientMultiton__cache) == 0
```

One of the nice things about the above pattern is that one can defer to the garbage collector for cleanup by simply removing any references to the object (in a Future for instance?)

Reply:
> I'm not sure what our appetite for metaclasses in the code base

Generally speaking I'm not a huge fan of metaclasses. Mostly because they are used very rarely and most people are having a hard time understanding it.

However, your suggestion looks pretty appealing and I'm intrigued. I'll look into this and would consider using the metaclass if it helps us to reduce code complexity in other areas.

Reply:
> We should keep this in mind. I think having this would help in some cases and would allow us to write a cleaner handlign of clients overall. I expect this to be a bigger effort than what I feel comfortable doing in this PR

There are also examples of a lazily initialised Multiton (with support for a finaliser) under a [BSD3 license](https://github.com/ratt-ru/dask-ms/blob/master/LICENSE) in the following locations

https://github.com/ratt-ru/dask-ms/blob/master/daskms/patterns.py
https://github.com/ratt-ru/dask-ms/blob/master/daskms/tests/test_patterns.py